reinforcement_learning/utils.py

```diff
@@ -15,4 +39,122 @@ def build_agenda_based_simulator(
     Returns:
         Agenda based user simulator.
     """
-    pass
+    domain = SimulationDomain(config["domain"].get())
+
+    item_collection = ItemCollection()
+    item_collection.load_items_csv(
+        config["items"].get(),
+        domain=domain,
+        domain_mapping=config["domain_mapping"].get(),
+        id_col=config["id_col"].get(),
+    )
+
+    ratings = Ratings(item_collection)
+    ratings.load_ratings_csv(file_path=config["ratings"].get())
+    historical_ratings, ground_truth_ratings = ratings.create_split(
+        config["historical_ratings_ratio"].get(0.8)
+    )
+
+    preference_model = SimplePreferenceModel(
+        domain,
+        item_collection,
+        historical_ratings,
+        historical_user_id="13",
+    )
+
+    # Loads dialogue sample
+    annotated_dialogues_file = config["dialogues"].get()
+    annotated_conversations = json_to_dialogues(
+        annotated_dialogues_file,
+    )
+
+    # Loads interaction model
+    interaction_model = InteractionModel(
+        config_file=config["intents"].get(),
+        annotated_conversations=annotated_conversations,
+    )
+
+    # NLU
+    nlu = get_NLU(config)
+
+    # NLG
+    template = extract_utterance_template(
+        annotated_dialogue_file=annotated_dialogues_file,
+    )
+    for i, v in template.items():
+        for utterance in v:
+            if utterance.text == "Do you have something more recent?":
+                v.remove(utterance)
+        template[i] = v
+
+    nlg = ConditionalNLG(template)
+
+    return UserSimulatorMovieBot(
+        config["simulator_id"].get(),
+        preference_model,
+        interaction_model,
+        nlu,
+        nlg,
+        domain,
+        item_collection,
+        ratings,
+    )
+
+
+def define_possible_actions(domain: MovieDomain) -> List[DialogueAct]:
+    """Defines the possible agent actions.
+
+    Args:
+        domain: The movie domain.
+    Returns:
+        List of possible actions as dialogue acts.
+    """
+    actions = []
+    actions += [
+        DialogueAct(
+            AgentIntents.ELICIT, [ItemConstraint(slot, Operator.EQ, "")]
+        )
+        for slot in domain.agent_requestable
+    ]
+    actions += [
+        DialogueAct(
+            AgentIntents.RECOMMEND,
+        ),
+        DialogueAct(
+            AgentIntents.CONTINUE_RECOMMENDATION,
+        ),
+    ]
+    actions.append(DialogueAct(AgentIntents.INFORM))
```

Comment:
Could you do it for all or at least some of these intents at once? It seems to be an easy function but it still takes a lot of lines of code.

Reply:
Done.